Add the ability to record any key/value pair in the 'custom' field in metadata.yaml

t0ny-peng opened the pull request:
Usage: `ros2 bag record --all --custom-data key1=value1 key2=value2`. The custom data will show up in the `metadata.yaml` file:
```yaml
rosbag2_bagfile_information:
  ...
  custom:
    key1: value1
    key2: value2
```

Closes [#547](https://github.com/ros2/rosbag2/issues/547)

MichaelOrlov commented on rosbag2_storage/include/rosbag2_storage/yaml.hpp:
```diff
@@ -41,6 +42,27 @@ void optional_assign(const Node & node, std::string field, T & assign_to)
   }
 }
 
+template<>
+struct convert<std::unordered_map<std::string, std::string>>
+{
+  static Node encode(const std::unordered_map<std::string, std::string> & custom_data)
+  {
+    Node node;
+    for (const auto & it : custom_data) {
+      node[it.first] = it.second;
+    }
+    return node;
+  }
+
+  static bool decode(const Node & node, std::unordered_map<std::string, std::string> & custom_data)
+  {
+    for (YAML::const_iterator it = node.begin(); it != node.end(); ++it) {
+      custom_data.emplace(it->first.as<std::string>(), it->second.as<std::string>());
+    }
+    return true;
+  }
+};
+
```
I think it would make sense to put this `convert` specialization to the `metadata_io.cpp` as others `convert` template specializations related to the metadata.

t0ny-peng replied:
Both `metadata_io.cpp` and `storage_options.cpp` needs this conversion so I placed it here

MichaelOrlov replied:
@t0ny-peng I don't really like that similar definitions spread out across the different files.  
It looks like that other `yaml convert` operators was incorrectly put in the `metadata_io.cpp` file from the very beginning.  
Since we have a dedicated file for `YAML` specific operations and definitions aka `yaml.hpp` it would more appropriate to all specific yaml conversions to this file.

Could you please move other `yaml` converter operators from  `metadata_io.cpp` to the `yaml.hpp` file?
i.e. everything under the `namespace YAML`.

I have tried it locally on my machine and it's compiles and passes unit tests.

t0ny-peng replied:
@MichaelOrlov OK moved here

MichaelOrlov commented on rosbag2_storage/include/rosbag2_storage/yaml.hpp:
```diff
@@ -41,6 +42,27 @@ void optional_assign(const Node & node, std::string field, T & assign_to)
+  static bool decode(const Node & node, std::unordered_map<std::string, std::string> & custom_data)
+  {
+    for (YAML::const_iterator it = node.begin(); it != node.end(); ++it) {
+      custom_data.emplace(it->first.as<std::string>(), it->second.as<std::string>());
```
`emplace` can fail to create new element. Need to add a sanity check for `emplace` result 
e.g.
```cpp
      if (!custom_data.emplace(it->first.as<std::string>(), it->second.as<std::string>()).second)
      {
        return false;
      }
```

t0ny-peng replied:
Since this method simply decode a yamlcpp Node to a `std::map`, I'm wondering in what case the `emplace` will fail? Note that the internal data structure in yamlcpp Node can already guarantee the uniqueness.

MichaelOrlov replied:
From https://en.cppreference.com/w/cpp/container/unordered_map/emplace
```cpp
template< class... Args >
std::pair<iterator,bool> emplace( Args&&... args );
```
> Return value
> Returns a pair consisting of an iterator to the inserted element, or the already-existing element if no insertion happened, and a bool denoting whether the insertion took place (true if insertion happened, false if it did not).

t0ny-peng replied:
Yes that is indeed the return value of `emplace` method. However, like I said, since we are decoding a YAML-cpp node to a STL map, the case of duplicate key, i.e., insertion failure, won't happen because YAML-cpp already did the deduplication.

MichaelOrlov commented on rosbag2_storage/src/rosbag2_storage/metadata_io.cpp:
```diff
@@ -225,6 +226,11 @@ struct convert<rosbag2_storage::BagMetadata>
         node["files"].as<std::vector<rosbag2_storage::FileInformation>>();
     }
+
+    if (node["custom"]) {
```
Out of curiosity, is it really a valid check `if (node["custom"])` ?
Will not it always return true as in case with std::unordered_map?

t0ny-peng replied:
`operator[]` of YAML-cpp Node will not create the element if the key doesn't exist.

MichaelOrlov commented on rosbag2_storage/src/rosbag2_storage/metadata_io.cpp:
```diff
@@ -199,6 +199,7 @@ struct convert<rosbag2_storage::BagMetadata>
     node["compression_mode"] = metadata.compression_mode;
     node["relative_file_paths"] = metadata.relative_file_paths;
     node["files"] = metadata.files;
+    node["custom"] = metadata.custom_data;
```
Suggestion to rename `custom` field to the `custom_data`.

t0ny-peng replied:
ok changed

MichaelOrlov reviewed (changes requested):
@t0ny-peng Thank you for contribution.  Overall looks good.
I have some minor findings. Please see them in inline comments.

I also have suggestion to rename `node["custom"]` to the `node["custom_data"]`.
name `custom`  is to generic and could confuse to what is relates for.

t0ny-peng commented on ros2bag/ros2bag/verb/record.py:
```diff
@@ -198,6 +205,13 @@ def main(self, *, args):  # noqa: D102
+        # Prepare custom_data dictionary
+        custom_data = {}
+        if args.custom_data:
+            key_value_pais = [pair.split('=') for pair in args.custom_data]
```
Fixed

t0ny-peng commented on ros2bag/ros2bag/verb/record.py:
```diff
@@ -198,6 +205,13 @@ def main(self, *, args):  # noqa: D102
+        if args.custom_data:
+            key_value_pais = [pair.split('=') for pair in args.custom_data]
+            custom_data = {pair[0]: pair[1] for pair in key_value_pais}
+            print(custom_data)
```
Added by mistake.  Removed

t0ny-peng commented on rosbag2_storage/include/rosbag2_storage/bag_metadata.hpp:
```diff
@@ -52,6 +53,7 @@ struct BagMetadata
   std::vector<TopicInformation> topics_with_message_count;
   std::string compression_format;
   std::string compression_mode;
+  std::unordered_map<std::string, std::string> custom_data;  // {key: value, ...}
```
Updated to 6.

t0ny-peng commented on ros2bag/ros2bag/verb/record.py:
```diff
@@ -148,6 +148,13 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
+            '--custom-data', type=str, metavar='KEY=VALUE', nargs='*',
+            help='Store the custom data in metadata.yaml'
+                 'under "rosbag2_bagfile_information/custom". The key=value pair can '
+                 'appear more than once.'
```
By appear more than once I mean the latter one will override the former ones. The description is updated.

MichaelOrlov reviewed (changes requested):
@t0ny-peng Thank you for the iteration in code review.
I have a few more requests for changes.
Also could you please next time push new changes as a separate commit?
One commit per one review iteration should be fine.
It's very difficult to review code outside of the Github e.g. in CLion, when all changes squashed in one commit and pushed with force command. i.e. there are no history for the changes outside of the PR.

Multiple commits will not clutter the final git history since we have mandatory squash for commits when we do merge on master. Although it's really helpful for code review.

t0ny-peng commented on rosbag2_storage/src/rosbag2_storage/metadata_io.cpp:
```diff
@@ -199,6 +199,9 @@ struct convert<rosbag2_storage::BagMetadata>
     node["files"] = metadata.files;
+    if (!metadata.custom_data.empty()) {
+      node["custom_data"] = metadata.custom_data;
+    }
```
Ok removed.

t0ny-peng commented:
@MichaelOrlov I've addressed your comments. I didn't split the first commit because it seems pretty atomic. The changes are all related and cannot be broken down to subcomponents. However moving the convert structs has its own commit. Please review. Thx

MichaelOrlov reviewed (approved):
@t0ny-peng Thanks, approving now.

As regards to the 
>  I didn't split the first commit because it seems pretty atomic. The changes are all related and cannot be broken down to subcomponents. However moving the convert structs has its own commit.

It's ok and this is exactly how I expected it to be.

MichaelOrlov commented:
Running CI
Gist: https://gist.githubusercontent.com/MichaelOrlov/f295a4dda3d8ce90076def0a0646c519/raw/620153ccb8176c963f45162d6b39e1903ee19bf2/ros2.repos
BUILD args: --packages-up-to ros2bag rosbag2_cpp rosbag2_storage rosbag2_tests
TEST args: --packages-select ros2bag rosbag2_cpp rosbag2_storage rosbag2_tests
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/10087/
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=16471)](http://ci.ros2.org/job/ci_linux/16471/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11061)](http://ci.ros2.org/job/ci_linux-aarch64/11061/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16825)](http://ci.ros2.org/job/ci_windows/16825/)

t0ny-peng commented:
@MichaelOrlov I'm not sure if increasing the version is the correct move here. `rosbag2_tests` has some database files that store version 5 at them, and the test failed because of version 6(from the code) and version 5(from the database file).


<img width="1124" alt="image" src="https://user-images.githubusercontent.com/4816327/160758216-4111df50-363a-4ee4-b5d3-5d6051bcd782.png">

In this case, what's the correct way of updating the database? Using sqlite to modify it manually?

MichaelOrlov commented:
@t0ny-peng I think we need to fix tests by updating databases.  
We need to change the metadata version number each time as we are changing something in the `BagMetadata` staruct.  
https://github.com/ros2/rosbag2/blob/master/rosbag2_storage/include/rosbag2_storage/bag_metadata.hpp#L44

Need to look to the failing tests, may be there are some routines for generating test database.

BTW I see that tests fails not only in `rosbag2_tests` also in [rosbag2_cpp](https://ci.ros2.org/job/ci_linux/16471/testReport/rosbag2_cpp/TemporaryDirectoryFixture/)

MichaelOrlov commented:
@t0ny-peng **Update:** Don't need to change database files.
I was able to fix failing tests from `rosbag2_tests` altering version number in https://github.com/t0ny-peng/rosbag2/blob/master/rosbag2_tests/resources/reindex_test_bags/target_metadata/multiple_files/metadata.yaml#L2 and adding new section for the `custom_data` at the end.

However I was not able to reproduce failures in [`rosbag2_cpp`](https://ci.ros2.org/job/ci_linux/16471/testReport/rosbag2_cpp/TemporaryDirectoryFixture/) 
with message:
> C++ exception with description "Exception on parsing info file: invalid node; first invalid key: "custom_data"" thrown in the test body.

t0ny-peng commented:
@MichaelOrlov I think the error in `rosbag2_cpp` has already been fixed. Could you please try the latest commit(aae2fc7b7069fe30205f09b92285ffc4ddb44a38)?

<img width="1193" alt="image" src="https://user-images.githubusercontent.com/4816327/160890598-d0690b06-6cda-472a-bb89-35a08597be66.png">

WRT changing the version in the test, I don't think comparing the current version of `metadata` CPP struct with the version in the test yaml file is the best way. It means the YAML file has to be updated every time the CPP metadata version increases. However for now I'll just do that as a one time patch.

MichaelOrlov commented:
Running CI again after fix in rosbag2_tests
Gist: https://gist.githubusercontent.com/MichaelOrlov/f295a4dda3d8ce90076def0a0646c519/raw/620153ccb8176c963f45162d6b39e1903ee19bf2/ros2.repos
BUILD args: --packages-up-to ros2bag rosbag2_cpp rosbag2_storage rosbag2_tests rosbag2_transport
TEST args: --packages-select ros2bag rosbag2_cpp rosbag2_storage rosbag2_tests rosbag2_transport
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/10089/
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=16473)](http://ci.ros2.org/job/ci_linux/16473/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11063)](http://ci.ros2.org/job/ci_linux-aarch64/11063/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16844)](http://ci.ros2.org/job/ci_windows/16844/)
